devicedefender/include/aws/iotdevicedefender/DeviceDefender.h

```diff
@@ -77,6 +77,148 @@ namespace Aws
+             *          ("metricName" or "metricFunc") is incorrect.
+             */
+            int RegisterCustomMetricIpAddressList(
+                aws_byte_cursor metricName,
+                aws_iotdevice_defender_get_ip_list_fn *metricFunc);
+
+            /**
+             * Registers a custom metric IP address list function to the Device Defender result. Will call the
+             * "metricFunc" function that is passed in each time a report is generated so it's data can be passed along
+             * with the other device defender payload data with the metric name of "metricName".
+             * @param metricName The key name for the data.
+             * @param metricFunc The function that is called to get the IP address list data.
+             * @return AWS_OP_SUCCESS if the custom metric was registered successfully.
+             *          Will return AWS_OP_ERR if the IP address list cannot be registered or the data passed
+             *          ("metricName" or "metricFunc") is incorrect.
+             */
+            int RegisterCustomMetricIpAddressList(
+                aws_byte_cursor metricName,
+                std::function<int(aws_array_list *, void *)> *metricFunc);
+
+            /**
+             * Registers a custom metric number that will report the CPU usage automatically into a custom metric called
+             * "cpu_usage". Calling this function will make the task report CPU usage each time a report is generated.
+             *
+             * Note: The CPU usage reported is in percentage ("12" = 12% CPU)
+             * @return AWS_OP_SUCCESS if the custom metric was registered successfully.
+             *      Will return AWS_OP_ERR if the CPU usage cannot be registered
+             */
+            int RegisterCustomMetricCpuUsage();
+
+            /**
+             * Registers a custom metric number that will report the RAM memory usage automatically into a custom metric
+             * called "memory_usage". Calling this function will make the task report memory usage each time a report is
+             * generated.
+             *
+             * Note: The memory usage reported is in kilobytes.
+             * @return AWS_OP_SUCCESS if the custom metric was registered successfully.
+             *      Will return AWS_OP_ERR if the CPU usage cannot be registered
+             */
+            int RegisterCustomMetricMemoryUsage();
+
+            /**
+             * Registers a custom metric number that will report the number of processors automatically into a custom
+             * metric called "processor_count". Calling this function will make the task report processor count each
+             * time a report is generated.
+             * @return AWS_OP_SUCCESS if the custom metric was registered successfully.
+             *      Will return AWS_OP_ERR if the CPU usage cannot be registered
+             */
+            int RegisterCustomMetricProcessorCount();
+
```

Comment:
These are technically not needed, they are just helper functions that use the new custom metrics functionality to add support for tracking these metrics.

By putting them here they make these metrics easy to add, but I'm not sure if we want them here or if instead we should have it just in the sample and point users to the sample to see how to implement tracking for these features.

On one hand, removing them would reduce the code and keep it lean, but on the other, it would require the user to implement this commonly requested functionality themselves. It is not too hard to add